contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,201 @@
+logic in one context or the other as needed.  last-appied and remote are not compared directly, and instead
+are only compared indirectly when the 2 diff results are merged.  Information that is redundant or
+should be checked for consistency across all 3 sources (e.g. checking for conflicts) is spread across
+3 logic locations - the first 2-way diff, the second 2-way diff and the merge of the 2 diffs.
+
+That the diffs each may produce multiple patch directives + results that constitute an update to a single
+field compounds the complexity of that comparing a single field occurs across 3 locations.
+
+Additionally, the logic to apply the patch must be updated both on the the server side
+and on the client, but cannot be done so transactionally.
+Because there there is no way for the client to know what directives a server supports (may be old),
+new directives must always be present but the patch must continue to function if they are ignored
+by the server.
+Because the server has no guarantee the client will send all directives it expects (may be old), it must
+not require any new directives to apply a patch "correctly".
+ 
+The diff / patch logic itself does not follow any sort of structure to encapsulate complexity
+into components so that logic doesn't bleed cross concerns.  The logic to collate the last-applied, local and
+remote field values, the logic to diff the field values and the logic to create the patch is
+all combined in the same group of package-scoped functions, instead of encapsulating
+each of these responsibilities in its own interface.
+
+There is no simple way to add a new patch directive (retain keys),
+to introduce customized logic for a specific field type (delete primitive from list)
+or to change the structure of a field (ordering list items)
+without touching dozens of locations across 1000+ lines of code where those locations
+are used in multiple contexts some of which are targeted by your changes and some of which
+are not.
+
+Lastly, sprinkling the implementation across dozens of locations makes it very challenging to
+flag guard the new behavior.  If issues are discovered during the stabilization period we cannot
+easily revert to the previous behavior by changing a default flag value.  The inability to build
+in these sorts of break-glass options further degrades confidence in safely accepting PRs.
+
+This is a text-book example of what the [Visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) 
+was designed to address.
+
+- Encapsulate logic in *Element*s and *Visitor*s
+- Introduce logic for new a field type by adding a new *Element* type
+- Introduce logic for new a merge strategy by defining a new *Visitor* implementation
+- Introduce logic on structuring of a field by updating the parsing function for that field type
+
+If the apply diff logic was redesigned, most of the preceding PRs could be implemented by
```

Comment:
Note that this would still change user-visible behavior. While technically bug fixes, we should consider how to communicate such changes to users, whether we need to retain prior behavior, and how we should transition to implementations with different behavior.

Reply:
Yes this makes sense.

In some cases the fixes purely fix something that was broken or unspecified (e.g. maintaining ordering for lists when merging is strictly more correct than randomizing the ordering).  In other cases I think it is pretty clearly an improvement, but the change should be communicated (e.g. deleting an item from a primitive lists now deletes it).  

The cases more related to your point would be fixing an unintended / broken merge-key or merge-strategy.